jpa/hibernate5_3/src/main/java/org/jboss/as/jpa/hibernate5/service/WildFlyCustomJtaPlatform.java

```diff
@@ -24,37 +24,69 @@
 import org.hibernate.engine.transaction.jta.platform.internal.JtaSynchronizationStrategy;
 
 /**
- * WildFlyCustomJtaPlatform
+ * WildFlyCustomJtaPlatform can obtain the JTA TransactionSynchronizationRegistry to be used by
+ * Hibernate ORM JPA + native applications.
+ * For JPA applications, we could of passed the TransactionSynchronizationRegistry into the
+ * constructor but Hibernate native apps wouldn't be able to do that, so this covers all app types.
  *
  * @author Scott Marlow
  */
-public class WildFlyCustomJtaPlatform extends JBossAppServerJtaPlatform {
+public class WildFlyCustomJtaPlatform extends JBossAppServerJtaPlatform implements JtaSynchronizationStrategy {
 
     private static final String TSR_NAME = "java:jboss/TransactionSynchronizationRegistry";
 
-    @Override
-    protected JtaSynchronizationStrategy getSynchronizationStrategy() {
-        final TransactionSynchronizationRegistry transactionSynchronizationRegistry =
-                locateTransactionSynchronizationRegistry();
+    // The 'transactionSynchronizationRegistry' used by JPA container managed applications,
+    // is reset every time the Transaction Manager service is restarted,
+    // as (application deployment) JPA persistence unit service depends on the TM service.
+    // For this reason, the static 'transactionSynchronizationRegistry' can be updated.
+    // Note that Hibernate native applications currently have to be (manually) restarted when the TM
+    // service restarts, as native applications do not have WildFly service dependencies set for them.
+    private static volatile TransactionSynchronizationRegistry transactionSynchronizationRegistry;
 
-        return new JtaSynchronizationStrategy() {
+    // JtaSynchronizationStrategy
+    @Override
+    public void registerSynchronization(Synchronization synchronization) {
+        transactionSynchronizationRegistry.registerInterposedSynchronization(synchronization);
+    }
 
+    // JtaSynchronizationStrategy
+    @Override
+    public boolean canRegisterSynchronization() {
+        return transactionSynchronizationRegistry.getTransactionStatus() == Status.STATUS_ACTIVE;
+    }
 
-            @Override
-            public void registerSynchronization(Synchronization synchronization) {
-                transactionSynchronizationRegistry.registerInterposedSynchronization(synchronization);
-            }
+    @Override
+    protected JtaSynchronizationStrategy getSynchronizationStrategy() {
+        return this;
+    }
 
-            @Override
-            public boolean canRegisterSynchronization() {
-                return transactionSynchronizationRegistry.getTransactionStatus() == Status.STATUS_ACTIVE;
+    private TransactionSynchronizationRegistry locateTransactionSynchronizationRegistry() {
```

Comment:
I'm not sure to understand the change correctly but it seems that now, code that was calling `locateTransactionSynchronizationRegistry()` will get the static `transactionSynchronizationRegistry` field that may come either from JNDI (as before) or from the new static `setTransactionSynchronizationRegistry` method, right?

Now that everyone is relying on this `transactionSynchronizationRegistry` field, could there be issues where JPA applications will get the one set by `setTransactionSynchronizationRegistry` instead of (another?) one from the JNDI lookup?

Should we distinguish the 2 cases (Hibenate native or JPA) instead of addressing both at the same time even though they rely on different mechanism to synchronise their states (JPA being able to use service dependencies)?

Reply:
> I'm not sure to understand the change correctly but it seems that now, code that was calling locateTransactionSynchronizationRegistry() will get the static transactionSynchronizationRegistry field that may come either from JNDI (as before) or from the new static setTransactionSynchronizationRegistry method, right?

Correct, we have two paths of initializing the static, which path is taken, depends on whether the first deployment is a native Hibernate or container managed JPA deployment.  The first deployment could also be both, it doesn't really matter which initialization path is taken, although JNDI is slower.  

> Now that everyone is relying on this transactionSynchronizationRegistry field, could there be issues where JPA applications will get the one set by setTransactionSynchronizationRegistry instead of (another?) one from the JNDI lookup?

No, as the transactionSynchronizationRegistry will always be the same value.  If the transaction manager service is restarted, all JPA applications will also restart and the transactionSynchronizationRegistry will be set to the new value to be used by all JPA applications.  For Hibernate native applications, they still have to be manually restarted (as in the past, which is an operational requirement that I have never heard a complaint about, likely that the TM service is never restarted for most users).

> Should we distinguish the 2 cases (Hibenate native or JPA) instead of addressing both at the same time even though they rely on different mechanism to synchronise their states (JPA being able to use service dependencies)?

We don't have a way to do anything else for Hibernate native.  

Thanks for the review, good to talk about ^ :)